docs/docs/data-model.md

```diff
@@ -0,0 +1,87 @@
+
+## Ordered Broadcast Channel
+
+The atomic unit of data in DriftDB is a **message**. Messages can be any non-`null` JSON value. The contents of messages are opaque to DriftDB’s operations.
+
+Every connection in DriftDB is scoped to a room. Messages sent over a connection are scoped to the same room as the connection.
+
+Each room corresponds to an ordered stream of messages. Each client will receive every message sent in a room while they are connected, in the same order.
+When the server rebroadcasts a message, it gives it a sequence number which represents its position in the room’s sequence. The first message in each room
+is given the sequence number 1.
+
+> (Why 1 and not 0? It’s common to need to store the sequence number of the “last message received”, and starting sequence numbers at 1 allows us to default that
+to 0 before any messages have been received.)
+
+## Keys
+
+Every message in DriftDB is associated with a **key**, which is a string. Any string is allowed, including the empty string.
+
+Keys can be used by an application to multiplex any number of “logical” streams on top of the global stream that DriftDB provides. Keys also matter to DriftDB
+for the purpose of replays.
+
+## Replays
+
+When a client connects to a room, DriftDB immediately includes the client in every broadcast to that room, but does not automatically replay old messages to
+the client.
+
+Clients can request a replay of messages sent to a given key, optionally starting from a particular sequence number.
+
+Internally, DriftDB stores messages in a hash map that associates each key with a growable ring buffer.
+
+## Actions
+
+Messages sent by clients are accompanied by an `action`, which determines whether they are added to the global broadcast stream, and whether they are replayable.
+
+DriftDB implements four actions.
+
+### `append`
+
+The `append` action results in a message being broadcast, and added to the replayable stream for its key.
+
+### `relay`
+
+The `relay` action results in a message being broadcast, but not added to the replay stream for its key. `relay` messages are still treated as part of the ordered
+broadcast and given a unique sequence number, but are discarded immediately by DriftDB immediately after they are broadcast.
+
+### `replace`
+
+The `replace` action broadcasts a message, and *replaces* the prior contents of the replay stream for just that message. You can think of `replace` as providing
+key-value semantics, where only the last value set for a given key is retained.
+
+### `compact`
+
+Unlike the other actions, `compact` must be accompanied by a sequence number. Also unlike the other actions, the compact action
+*does not* broadcast the message. Instead, it deletes all messages from the replay stream of the given key, up to and including the given sequence number.
+
+`compact` then inserts the accompanying message at the *bacK* of the replay stream for the given key, with the sequence number provided.
```

Comment:
This sentence is a little confusing to me. Maybe "the back of the replay stream" is ambiguous?